enhancements/installer/custom-dns.md

```diff
@@ -0,0 +1,574 @@
+need to be resolvable. So, no additional changes are needed during the bootstrap
+phase to support this feature.
+
+10. The MCO is also responsible for generating manifests for running CoreDNS on
+the master and worker nodes using information provided in the Infrastructure CR.
+
+### API Extensions
+
+#### api.dns [https://github.com/openshift/api/blob/master/config/v1/types_dns.go]
+
+We need a way to inform the Ingress Controller to not configure the DNS records
+for the Ingress service in the cloud DNS when the custom DNS has already been
+configured. So, adding a string config to the DNSSpec of the DNS CR to
+indicate whether the DNS configuration has already been handled externally
+should take care of that.
+
+```go
+
+type DNSSpec struct {
+// <snip>
+
+// ingressDNSType indicates whether the DNS configuration for the Ingress
+// service has been handled externally by a custom DNS solution.
+// A value of "External" indicates that the Ingress Controller does not have
+// to configure the cloud DNS and a default value of "Internal" indicates
+// that the Ingress Controller is responsible for the DNS configuration.
+
+// Once set, this field cannot be changed.
+// +optional
+IngressDNSType string `json:"ingressDNSType"`
+}
+```
+
+#### Infrastructure CR [https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go]
+
+```go
+
+type AWSPlatformStatus struct {
+// <snip>
+
+// A field with the name APIServerInternalIPs is used in the BareMetal platform to
+// hold 2 IP addresses for the dual-stack case.
+// In this case, we want to pass the IP addresses the API Server resolves to via
+// this field. At this time, the understanding is that it could me multiple (> 2)
+// IP addresses even in non-dualstack case.
+
+// Once set, this field cannot be changed.
+// +optional
+// Note: Looking for a better name for this field
+CustomAPIServerInternalIPs []string `json:"customApiServerInternalIPs"`
+}
+```
+
+#### AWS Platform Struct within the Install Config [https://github.com/openshift/installer/blob/master/pkg/types/aws/platform.go]
+
+```go
+
+type CustomDNS struct {
+    // customerManagedDNS is a string that needs to be explicitly set to
+    // "Enabled" when customer does not want the Installer to configure
+    // Route53 for them. The customer is responsible for responsible for
+    // configuring their own DNS solution based on LB configuration geberated
+    // by the Installer.
+    // Default value for this field is "Disabled". When "Disabled" the other
+    // fields in this struct will not be validated or used. 
+   
+    CustomerManagedDNS string `json:"customerManagedDNS"`
+
+    // customAPIServerInternalIPs is a list of IP addresses for the internal
+    // API server. There should be one IP address per subnet provided in the
+    // AWS platform config. These IP addresses should be validated to be within
+    // the machine network.
+    CustomAPIServerInternalIPs []string `json:"customAPIServerInternalIPs"`
```

Comment:
Ideally I hope we would be able to get away without having the users pass in specific internal IP addresses, either by grabbing them from passed in subnets or the machine network cidr.

Reply:
The user facing config for IP addresses is optional. Let me update the enhancement with more information on what the Installer can do when not provided.